velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,226 @@
+        } else {
+          splitInner(
+              strings->valueAt<StringView>(row),
+              delim,
+              limit,
+              row,
+              resultWriter);
+        }
+      });
+    }
 
     resultWriter.finish();
-
-    // Reference the input StringBuffers since we did not deep copy above.
+    // Ensure that our result elements vector uses the same string buffer as
+    // the input vector of strings.
     result->as<ArrayVector>()
         ->elements()
         ->as<FlatVector<StringView>>()
-        ->acquireSharedStringBuffers(args[0].get());
+        ->acquireSharedStringBuffers(strings->base());
   }
 
  private:
-  const char pattern_;
-};
-
-/// This class will be updated in the future as we support more variants of
-/// split
-class Split final : public exec::VectorFunction {
- public:
-  Split() {}
+  // When pattern is empty, split each character out. Since Spark 3.4, when
+  // delimiter is empty, the result does not include an empty tail string, e.g.
+  // split('abc', '') outputs ["a", "b", "c"] instead of ["a", "b", "c", ""].
+  // The result does not include remaining string when limit is smaller than the
+  // string size, e.g. split('abc', '', 2) outputs ["a", "b"] instead of ["a",
+  // "bc"].
+  void splitEmptyDelimiter(
+      const StringView current,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+    if (current.size() == 0) {
+      arrayWriter.add_item().setNoCopy(StringView());
+      resultWriter.commit();
+      return;
+    }
+
+    const char* const begin = current.begin();
+    const char* const end = current.end();
+    const char* pos = begin;
+    while (pos < end && pos < limit + begin) {
+      arrayWriter.add_item().setNoCopy(StringView(pos, 1));
+      pos += 1;
+    }
+    resultWriter.commit();
+  }
 
-  void apply(
-      const SelectivityVector& rows,
-      std::vector<VectorPtr>& args,
-      const TypePtr& /* outputType */,
-      exec::EvalCtx& context,
-      VectorPtr& result) const override {
-    auto delimiterVector = args[1]->as<ConstantVector<StringView>>();
-    VELOX_CHECK(
-        delimiterVector, "Split function supports only constant delimiter");
-    auto patternString = args[1]->as<ConstantVector<StringView>>()->valueAt(0);
-    VELOX_CHECK_EQ(
-        patternString.size(),
-        1,
-        "split only supports only single-character pattern");
-    char pattern = patternString.data()[0];
-    SplitCharacter splitCharacter(pattern);
-    splitCharacter.apply(rows, args, nullptr, context, result);
+  // Split with a non-empty pattern.
+  void splitInner(
+      StringView input,
+      const StringView delim,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    VELOX_CHECK(!delim.empty(), "Delimiter size should not empty");
```

Comment:
Use VELOX_USER_CHECK instead. The throw of VELOX_CHECK indicates a bug of Velox, while the throw of VELOX_USER_CHECK indicates that the exception is caused by user.

Delimiter size should not empty -> Non-empty delimiter is expected.